src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/NuGetRemoteFileService.cs

```diff
@@ -0,0 +1,264 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+#nullable enable
+
+using System;
+using System.Collections.Specialized;
+using System.Diagnostics;
+using System.IO;
+using System.Net.Http;
+using System.Runtime.Caching;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft;
+using Microsoft.ServiceHub.Framework;
+using Microsoft.ServiceHub.Framework.Services;
+using NuGet.Common;
+using NuGet.PackageManagement.Telemetry;
+using NuGet.Packaging;
+using NuGet.Packaging.Core;
+using NuGet.VisualStudio.Internal.Contracts;
+using NuGet.VisualStudio.Telemetry;
+
+namespace NuGet.PackageManagement.VisualStudio
+{
+    public sealed class NuGetRemoteFileService : INuGetRemoteFileService
+    {
+        private ServiceActivationOptions? _options;
+        private IServiceBroker _serviceBroker;
+        private AuthorizationServiceClient? _authorizationServiceClient;
+        private INuGetTelemetryProvider _nuGetTelemetryProvider;
+        private bool _disposedValue;
+
+        internal readonly static MemoryCache IdentityToUriCache = new MemoryCache("PackageSearchMetadata",
+            new NameValueCollection
+            {
+                { "cacheMemoryLimitMegabytes", "4" },
+                { "physicalMemoryLimitPercentage", "0" },
+                { "pollingInterval", "00:02:00" }
+            });
+
+        private static readonly CacheItemPolicy CacheItemPolicy = new CacheItemPolicy
+        {
+            SlidingExpiration = ObjectCache.NoSlidingExpiration,
+            AbsoluteExpiration = ObjectCache.InfiniteAbsoluteExpiration,
+        };
+
+        public static void AddIconToCache(PackageIdentity packageIdentity, Uri iconUri)
+        {
+            string key = "icon:" + packageIdentity.ToString();
+            if (iconUri != null && !IdentityToUriCache.Add(key, iconUri, CacheItemPolicy))
+            {
+                IdentityToUriCache.Remove(key);
+                IdentityToUriCache.Add(key, iconUri, CacheItemPolicy);
+            }
+        }
+
+        public static void AddLicenseToCache(PackageIdentity packageIdentity, Uri embeddedLicenseUri)
+        {
+            Assumes.NotNull(embeddedLicenseUri);
+            string key = "license:" + packageIdentity.ToString();
+            if (!IdentityToUriCache.Add(key, embeddedLicenseUri, CacheItemPolicy))
+            {
+                IdentityToUriCache.Remove(key);
+                IdentityToUriCache.Add(key, embeddedLicenseUri, CacheItemPolicy);
+            }
+        }
```

Comment:
I guess this is fine for now, but I'd really like it if now-or-eventually, this is abstracted away into a set of supported "Well Known Files"--for example, if we want to add Changelog or docs support in the future.

Reply:
good question. should we build this into the api now...worth considering so we don't have to version it later. let me think.

Reply:
thought a bit about it... i'd imagine we could pass an enum into a method... packageIdentity, wellKnownEmbeddedFile

where wellKnownEmbeddedFiles might be readme, license and what else?
maybe icon...but that isn't always embedded, can also be a url outside of the package.
i lean towards not doing this in a generic way yet...as we don't know the full set of needs yet...and since the nuspec is on the server, we cannot just take a relative path, etc...

would love to discuss, hear other thoughts.

Reply:
added issue to spec: https://github.com/NuGet/Home/pull/10460
closing now, cause i don't think we'll do now...but worth discussing.